coderd/provisionerjobs.go

```diff
@@ -343,3 +338,46 @@ func convertProvisionerJob(provisionerJob database.ProvisionerJob) codersdk.Prov
 func provisionerJobLogsChannel(jobID uuid.UUID) string {
 	return fmt.Sprintf("provisioner-log-logs:%s", jobID)
 }
+
+// provisionerJobLogsMessage is the message type published on the provisionerJobLogsChannel() channel
+type provisionerJobLogsMessage struct {
+	EndOfLogs bool                         `json:"end_of_logs,omitempty"`
+	Logs      []database.ProvisionerJobLog `json:"logs,omitempty"`
+}
+
+func (api *API) followLogs(jobID uuid.UUID) (chan database.ProvisionerJobLog, func(), error) {
+	bufferedLogs := make(chan database.ProvisionerJobLog, 128)
+	closeSubscribe, err := api.Pubsub.Subscribe(provisionerJobLogsChannel(jobID),
+		func(ctx context.Context, message []byte) {
+			jlMsg := provisionerJobLogsMessage{}
+			err := json.Unmarshal(message, &jlMsg)
+			if err != nil {
+				api.Logger.Warn(ctx,
+					fmt.Sprintf("invalid provisioner job log on channel %q: %s",
+						provisionerJobLogsChannel(jobID),
+						err.Error()))
+				return
+			}
+
+			for _, log := range jlMsg.Logs {
+				select {
+				case bufferedLogs <- log:
+					api.Logger.Debug(ctx, "subscribe buffered log",
+						slog.F("job_id", jobID),
+						slog.F("stage", log.Stage))
+				default:
+					// If this overflows users could miss logs streaming. This can happen
+					// if a database request takes a long amount of time, and we get a lot of logs.
+					api.Logger.Warn(ctx, "provisioner job log overflowing channel")
```

Comment:
I don't totally get this comment. Doesn't the messages from from the database? So doesn't this overflow if the caller (websocket) is slower than the logs coming in over the database? 


Just looking for clarity of the comment, I know we should not block the database side of this. I wonder if we should record the number of skipped logs and try to insert something like "... X logs lines skipped ...".

Reply:
Yeah, the channel could overflow if the client is slow to read from the websocket, but it could also overflow if more than 128 messages are added between the time we set up the pubsub subscription and the time we start reading from the channel, which doesn't happen until `GetProvisionerLogsByIDBetween` completes.

Reply:
Yep, yep.

The problem with trying to insert a  "... X logs lines skipped ..." message is that the buffer is already full, so we don't have anywhere to put it.  We'd have to roll our own ring buffer or something messy to be able to coalesce logs already buffered.  I don't want to do that in this PR --- this is just a refactor matching existing behavior.